### ISSUE TITLE
Longtable continuation header on first page

### ISSUE THREAD

**Issue by paulQdata** [latex3/latex2e#672]
## Brief outline of the bug
I have document containg multiples longtables that use the \endfirsthead and \endhead commands to show headers on the first page/second page of the table. Seemingly unrelated changes to the structure of the table (font size) or margins have the effect of showing the continuations headers on the FIRST page of the table.  Is there something wrong with the structure of this table?

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}       % <--should be always the first line (see CONTRIBUTING)!
\documentclass{article}
\usepackage{longtable}

  % Any preamble code goes here

[example.log](https://github.com/latex3/latex2e/files/7202511/example.log)
[example.pdf](https://github.com/latex3/latex2e/files/7202512/example.pdf)
\begin{document}
\vspace*{\fill}
Filler
\section{Section}
\subsection{Subsection}
\begin{longtable}[l]{l l l}
F1 & F2 & F3
\endfirsthead
S1 & S2 & S3
\endhead
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\
1 & 2 & 3 \\

\
[example.pdf](https://github.com/latex3/latex2e/files/7202534/example.pdf)
end{longtable}
\newpage
%%

  % Demonstration of issue here
  
\end{document}
```

**Comment by davidcarlisle** [comment]
You didn't mention it but the initial head is being output but at the top of the page, above the section heading

![image](https://user-images.githubusercontent.com/1268738/134160449-73642e22-b8a9-4dec-91a8-e1973e8866cd.png)

I assume you see the same?

I have over the years seen some misplaced headers at the end of the table, but I've never seen this which seems to be a new bug.

So thanks for the test file, I'll need to trace what it is doing (basically the fill glue is confusing longtable, it assumes the section heading will be output on the previous page and things go wrong...

An immediate workaround would be to put `\clearpage` before the `\section` so that the page break is handled before longtable starts, but that is not a real solution.

**Comment by paulQdata** [comment]
Yes. The initial head appears above the section header for me too. I also have a rather more complex example without any \fill glue. I will try and see if I can cut it down to something manageable.

**Comment by paulQdata** [comment]
This is another example in which the same fault occurs. My tex file consists of a number of sections, each containing one longtable. Preceding each section is the command \needspace{6cm}, needed because there is otherwise an tendency for the section title to end up at the bottom of the page.  

\documentclass[a4paper]{article}
\usepackage{longtable}
\usepackage{needspace}
\begin{document}
\subsection{A}
\begin{longtable}[l]{l}
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
1 \\
\end{longtable}
\needspace{6cm}
\subsection{B}
\begin{longtable}[l]{l l l }
\rule{3cm}{4cm}\\
\endfirsthead
BS1 & \\
\endhead
1 \\*
1 \\
1 \\*
1 \\
1 \\*
1 \\
1 \\*
1 \\
1 \\*
1 \\
1 \\*
1 \\
1 \\*
\end{longtable}
\end{document}


[needspace_example.pdf](https://github.com/latex3/latex2e/files/7242644/needspace_example.pdf)

[needspace_example.log](https://github.com/latex3/latex2e/files/7242645/needspace_example.log)

**Comment by josephwright** [comment]
Another one for v5?
